packages/auth/lib/Users.js

```diff
@@ -0,0 +1,372 @@
+        console.error('multiple different session?!')
+        throw new SessionTokenValidationFailed({ email: emailFromQuery })
+      }
+      session = curSession
+    } else if (!session) {
+      console.error('session is required to validate against')
+      throw new SessionTokenValidationFailed({ email: emailFromQuery })
+    }
+
+    const validated = await validateChallenge({ pgdb, session, user: existingUser, ...tokenChallenge })
+    if (!validated) {
+      console.error('wrong token')
+      throw new SessionTokenValidationFailed({ email: emailFromQuery, ...tokenChallenge })
+    }
+    tokenTypes.push(tokenChallenge.type)
+  }
+
+  // security net
+  if (tokenTypes.length < 2 && (existingUser && existingUser.isTwoFactorEnabled)) {
+    console.error('two factor is enabled but less than 2 challenges provided')
+    throw new SessionTokenValidationFailed({ email: emailFromQuery })
+  }
+
+  // verify and/or create the user
+  const { user, isVerificationUpdated } = await upsertUserVerified({
+    pgdb,
+    email: session.sess.email
+  })
+
+  const transaction = await pgdb.transactionBegin()
+  try {
+    // log in the session and delete token
+    await transaction.public.sessions.updateOne({
+      id: session.id
+    }, {
+      sess: {
+        ...session.sess,
+        passport: {
+          user: user.id
+        }
+      }
+    })
+
+    // let the tokens expire
+    await transaction.public.tokens.update({
+      sessionId: session.id
+    }, {
+      updatedAt: new Date(),
+      expiresAt: new Date()
+    })
+    transaction.transactionCommit()
+  } catch (error) {
+    transaction.transactionRollback()
+    throw new AuthorizationFailedError({ session: session })
+  }
+
+  // call signIn hooks
+  try {
+    await Promise.all(
+      signInHooks.map(hook =>
+        hook(
+          user.id,
+          isVerificationUpdated,
+          pgdb
+        )
+      )
+    )
+  } catch (e) {
+    console.warn(`sign in hook failed in authorizeSession`, e)
+  }
+  return user
+}
+
+const upsertUserVerified = async({ pgdb, email }) => {
+  const transaction = await pgdb.transactionBegin()
+  try {
+    const existingUser = await transaction.public.users.findOne({ email })
+    const user = existingUser ||
+      await transaction.public.users.insertAndGet({
+        email,
+        verified: true
+      })
+    if (!user.verified) {
+      await transaction.public.users.updateOne({
+        id: user.id
+      }, {
+        verified: true
+      })
+    }
+    await transaction.transactionCommit()
+    return {
+      user,
+      isVerificationUpdated: (!existingUser || !existingUser.verified)
+    }
+  } catch (error) {
+    await transaction.transactionRollback()
+    throw error
+  }
+}
+
+const resolveUser = async ({ slug, pgdb, fallback }) => {
+  const user = await pgdb.public.users.findOne(
+  isUUID.v4(slug)
+    ? {id: slug}
+    : {username: slug})
+  return user || fallback
+}
```

Comment:
remove fallback, this is not the intended behaviour. If one calls `updateEmail`,  `clearSession`, etc. and providing a  `userId` for a user which doesn't exist, the mutation must throw and not just assume me as the user. This could lead to unintended changes on me, when I would want to change another user. see [updateEmail](https://github.com/orbiting/backends/pull/2/files#diff-deae4cda0bc3ec807b15232f18604798L22)